larray/core/session.py

```diff
@@ -1475,6 +1492,199 @@ def display(k, v, is_metadata=False):
+    ...     mortality_rate: ConstantAxesArray((AGE, GENDER))
+    ...     # --- declare an array with constant axes  (with default value) ---
+    ...     mortality: ConstantAxesArray((AGE, GENDER, TIME)) = 0
+
+    >>> def run_model(variant_name):
+    ...     # instantiation --> create an instance of the ModelVariables class
+    ...     m = ModelVariables(
+    ...                       variant_name = variant_name,
+    ...                       birth_rate = zeros((AGE, GENDER)),
+    ...                       births = zeros((AGE, GENDER, TIME)),
+    ...                       mortality_rate = 0,
+    ...                       )
+    ...     # ==== model ====
+    ...     # some code here
+    ...     # ...
+    ...     # axes of arrays 'birth_rate' and 'births' are not protected
+    ...     m.birth_rate = full((AGE, GENDER, TIME), fill_value=0.05)
+    ...     # axes of arrays 'mortality_rate' and 'mortality' are protected (and known)
+    ...     m.mortality_rate = full((AGE, GENDER), fill_value=0.05)
+    ...     m.mortality = 0
+    ...     # it still possible to add undeclared variables but this must be done with caution.
+    ...     m.undeclared_var = 'undeclared_var'
+    ...     # ...
+    ...     # ==== output ====
+    ...     # save all variables in an HDF5 file
+    ...     m.save('{}.h5'.format(variant_name), display=True)
+
+    Content of file 'main.py'
+
+    >>> run_model('variant_1')
+    dumping variant_name ... done
+    dumping birth_rate ... done
+    dumping births ... done
+    dumping mortality_rate ... done
+    dumping mortality ... done
+    dumping population ... done
+    dumping undeclared_var ... done
+    """
+    class Config:
+        # whether to allow arbitrary user types for fields (they are validated simply by checking
+        # if the value is an instance of the type). If False, RuntimeError will be raised on model declaration
+        # (default: False)
+        arbitrary_types_allowed = True
+        # whether to validate field defaults
+        validate_all = True
+        # whether to ignore, allow, or forbid extra attributes during model initialization (and after).
+        # Accepts the string values of 'ignore', 'allow', or 'forbid',
+        # or values of the Extra enum (default: Extra.ignore)
+        extra = Extra.allow
+        # whether to perform validation on assignment to attributes
+        validate_assignment = True
+        # whether or not models are faux-immutable, i.e. whether __setattr__ is allowed.
+        # (default: True)
+        allow_mutation = True
+
+    def __init__(self, *args, **kwargs):
+        meta = kwargs.pop('meta', Metadata())
+
+        # create an intermediate Session object to not call the __setattr__
+        # and __setitem__ overridden in the present class
+        s = Session(*args, **kwargs)
+
+        # Warning: order of fields is not preserved (seems like fields with default values comes after)
+        items = dict(s.items())
```

Comment:
FWIW, as far as the **items below are concerned, you do not need to convert s to a dict, you can use **s directly.
But maybe that does not work for the `items.update(self.__field_defaults__)` line?